gix-index/src/entry/stat.rs

```diff
@@ -0,0 +1,415 @@
+        Time {
+            secs: value.unix_seconds().try_into().expect("can't represent non-unix times"),
+            nsecs: value.nanoseconds(),
+        }
+    }
+}
+
+impl PartialEq<FileTime> for Time {
+    fn eq(&self, other: &FileTime) -> bool {
+        *self == Time::from(*other)
+    }
+}
+
+impl PartialOrd<FileTime> for Time {
+    fn partial_cmp(&self, other: &FileTime) -> Option<Ordering> {
+        self.partial_cmp(&Time::from(*other))
+    }
+}
+
+/// Configuration for comparing stat entries
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+pub struct Options {
+    /// If true, a files creation time is taken into consideration when checking if a file changed.
+    /// Can be set to false in case other tools alter the creation time in ways that interfere with our operation.
+    ///
+    /// Default true.
+    pub trust_ctime: bool,
+    /// If true, all stat fields will be used when checking for up-to-date'ness of the entry. Otherwise
+    /// nano-second parts of mtime and ctime,uid, gid, inode and device number _will not_ be used, leaving only
+    /// the whole-second part of ctime and mtime and the file size to be checked.
+    ///
+    /// Default true.
+    pub check_stat: bool,
+    // TODO: enable by default? Seems to be disabled in git??
+    // documentation only talks about this being disabled for linux but I can't
+    // see it being enabled on other OS
+    /// Whether to compare nano secs when comparing timestamps. This currently
+    /// leads to many false positives on linux and is therefore disabled there.
+    ///
+    /// Default false
+    pub use_nsec: bool,
+    /// Whether to compare network devices secs when comparing timestamps.
+    /// Disabled by default because this can cause many false positives on network
+    /// devices where the device number is not stable
+    ///
+    /// Default false
+    pub use_stdev: bool,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            trust_ctime: true,
+            check_stat: true,
+            use_nsec: false,
+            use_stdev: false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use filetime::FileTime;
+
+    use crate::entry::stat::{Options, Time};
+    use crate::entry::Stat;
+
+    #[test]
+    fn use_nsec() {
+        let stat1 = Stat {
+            mtime: Time { secs: 0, nsecs: 0 },
+            ctime: Time { secs: 0, nsecs: 0 },
+            dev: 0,
+            ino: 0,
+            uid: 0,
+            gid: 0,
+            size: 0,
+        };
+        let stat2 = Stat {
+            mtime: Time { secs: 0, nsecs: 10 },
+            ctime: Time { secs: 0, nsecs: 0 },
+            dev: 0,
+            ino: 0,
+            uid: 0,
+            gid: 0,
+            size: 0,
+        };
+
+        assert!(stat1.matches(&stat2, Options::default()));
+        assert!(!stat1.matches(
+            &stat2,
+            Options {
+                use_nsec: true,
+                ..Default::default()
```

Comment:
I always love to see the `msg` portion of `assert!` to be used to spell things out in plain text.

Reply:
I added comments in [5edcc13](https://github.com/Byron/gitoxide/pull/805/commits/5edcc1399519611ec6380bb1ca6ad66072fa0b64).